[WGSL] Use String instead of StringView for Identifiers

tadeuzagallo opened the pull request:
#### b965bbcd761815ab412e357c784406a321283f77
<pre>
[WGSL] Use String instead of StringView for Identifiers
<a href="https://bugs.webkit.org/show_bug.cgi?id=250680">https://bugs.webkit.org/show_bug.cgi?id=250680</a>
&lt;rdar://problem/104299588&gt;

Reviewed by Myles C. Maxfield.

There should be no extra cost if we create the string without copying and it
allows us to use runtime-generated strings which will be necessary soon.

* Source/WebGPU/WGSL/AST/ASTArrayAccess.h:
* Source/WebGPU/WGSL/AST/ASTAttribute.h:
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h:
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTGlobalDirective.h:
(WGSL::AST::GlobalDirective::GlobalDirective):
(WGSL::AST::GlobalDirective::name const):
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTLocationAttribute.h:
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::ParameterizedType::stringToKind):
(WGSL::AST::ParameterizedType::stringViewToKind): Deleted.
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer&lt;T&gt;::lex):
* Source/WebGPU/WGSL/Lexer.h:
(WGSL::Lexer::makeIdentifierToken):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseTypeDecl):
(WGSL::Parser&lt;Lexer&gt;::parseTypeDeclAfterIdentifier):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Token.h:
(WGSL::Token::Token):
(WGSL::Token::operator=):
(WGSL::Token::~Token):

Canonical link: <a href="https://commits.webkit.org/259044@main">https://commits.webkit.org/259044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6faee3c00c334c83c6b0885e84b33d60eb571663

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112945 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3728 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95972 "Failed to checkout and rebase branch from PR 8697") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112080 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109489 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93745 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/95972 "Failed to checkout and rebase branch from PR 8697") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92508 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25351 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/95972 "Failed to checkout and rebase branch from PR 8697") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6195 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26740 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3270 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46266 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6220 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8129 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h:
```diff
@@ -33,17 +33,17 @@ class BuiltinAttribute final : public Attribute {
 
 public:
-    BuiltinAttribute(SourceSpan span, StringView name)
+    BuiltinAttribute(SourceSpan span, String name)
```
You're doing a double copy here; surely you want the parameter to be `const String&`? Or you could keep the parameter a `StringView` but still change the parameter to `String m_name` like you're doing below.

tadeuzagallo replied:
That's fair, I didn't worry too much since the underlying string isn't being copied anyway, so it's equivalent to the StringView code (right?). But in any case, no downside in making it `const String&`.

litherum commented on Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
```diff
@@ -66,7 +66,7 @@ class FunctionDecl final : public Decl {
     using List = UniqueRefVector<FunctionDecl>;
 
-    FunctionDecl(SourceSpan sourceSpan, StringView name, Parameter::List&& parameters, std::unique_ptr<TypeDecl>&& returnType, CompoundStatement&& body, Attribute::List&& attributes, Attribute::List&& returnAttributes)
+    FunctionDecl(SourceSpan sourceSpan, String name, Parameter::List&& parameters, std::unique_ptr<TypeDecl>&& returnType, CompoundStatement&& body, Attribute::List&& attributes, Attribute::List&& returnAttributes)
```
Ditto

litherum commented on Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
```diff
@@ -27,25 +27,25 @@
 
 public:
-    IdentifierExpression(SourceSpan span, StringView identifier)
+    IdentifierExpression(SourceSpan span, String identifier)
```
Ditto, and ditto elsewhere too.

djg reviewed (approved):
:shipit: with Myles nit fixed.

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
```diff
@@ -41,7 +41,7 @@ class Parameter final : public Node {
-    Parameter(SourceSpan span, StringView name, UniqueRef<TypeDecl>&& type, Attribute::List&& attributes)
+    Parameter(SourceSpan span, const String& name, UniqueRef<TypeDecl>&& type, Attribute::List&& attributes)
         : Node(span)
         , m_name(WTFMove(name))
```
Oops, forgot this `WTFMove` here and now `name` is a constant. So weird that local builds aren't failing though...

tadeuzagallo committed:
[WGSL] Use String instead of StringView for Identifiers
https://bugs.webkit.org/show_bug.cgi?id=250680
<rdar://problem/104299588>

Reviewed by Myles C. Maxfield.

There should be no extra cost if we create the string without copying and it
allows us to use runtime-generated strings which will be necessary soon.

* Source/WebGPU/WGSL/AST/ASTArrayAccess.h:
* Source/WebGPU/WGSL/AST/ASTAttribute.h:
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h:
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTGlobalDirective.h:
(WGSL::AST::GlobalDirective::GlobalDirective):
(WGSL::AST::GlobalDirective::name const):
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTLocationAttribute.h:
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::ParameterizedType::stringToKind):
(WGSL::AST::ParameterizedType::stringViewToKind): Deleted.
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer<T>::lex):
* Source/WebGPU/WGSL/Lexer.h:
(WGSL::Lexer::makeIdentifierToken):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseTypeDecl):
(WGSL::Parser<Lexer>::parseTypeDeclAfterIdentifier):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Token.h:
(WGSL::Token::Token):
(WGSL::Token::operator=):
(WGSL::Token::~Token):

Canonical link: https://commits.webkit.org/259044@main
